Remove most unused includes

BenWiederhake opened the pull request:
This removes thousands of useless and unjustified `#include`s in over 1000 files, resulting in a speed up:

```
$ ninja clean && CCACHE_DISABLE=1 /usr/bin/time ninja -j16 # before
[0/2] Re-checking globbed directories...
[2/2] Cleaning all built files...
Cleaning... 4126 files.
[0/2] Re-checking globbed directories...
[4122/4122] Linking CXX executable Tests/LibWeb/test-web
2810.56user 163.99system 3:22.81elapsed 1466%CPU (0avgtext+0avgdata 493584maxresident)k
0inputs+4423816outputs (3707major+59077981minor)pagefaults 0swaps
$ ninja clean && CCACHE_DISABLE=1 /usr/bin/time ninja -j16 # after
[0/2] Re-checking globbed directories...
[2/2] Cleaning all built files...
Cleaning... 4129 files.
[0/2] Re-checking globbed directories...
[4122/4122] Linking CXX executable Tests/LibWasm/test-wasm
2738.57user 154.10system 3:17.84elapsed 1462%CPU (0avgtext+0avgdata 492068maxresident)k
0inputs+4420504outputs (3903major+56631274minor)pagefaults 0swaps
```

A reduction of 80 seconds time in userspace (or about 2.8%). Not overwhelming, but hopefully enough to make you click "merge".

Given how quickly Serenity changes, this will interfere with every single PR. I'm sorry.

For every removed `#include`, my script did the following:
- Determine the main things (classes, enums, functions, structs, etc) defined in the includ**ed** file
- Check that none of them are used in the includ**ing** file
- Check that Serenity builds cleanly before and after removing that line.

Given the sheer amount of removed includes, I'm sure that this will break something, so this needs to be bisectable. I tried to choose reasonable boundaries, resulting in about 140 commits for about 1000 touched files, ranging from 3 to 30 files in each commit. I think that's reasonable. If you disagree, please tell me how else I should group them.

For shits and giggles, here's a tag with one file per commit: https://github.com/BenWiederhake/serenity/commits/historic/too-many-commits

The script itself is far from ready to be shared, but this "first step" of `#include` removals should be good enough to warrant being merged already.

BenWiederhake commented:
Changes since last push:
- Rebase to avoid conflict
- Fix Kernel/Arch breakage by removing those commits (I'll look at it later in a different PR)
- Assign authorship to `Ben Wiederhake <see other commits>` to avoid accusations that I'm doing this only to enter a "high-number commit club"

bgianfo commented:
> The script itself is far from ready to be shared, but this "first step" of `#include` removals should be good enough to warrant being merged already.

I think if it's "ready" enough to send out hundreds of commits from, it should be be ready enough to share with people so we can scrutinize it's methodology to help validate these changes.  :smile:

BenWiederhake commented:
The script I ran can be found here: https://gist.github.com/BenWiederhake/727709ee95809eb60f2944fca30e2775

Afterwards, I did some manual editing, especially in AK/, when I saw that an include really was unjustified, but wouldn't build without it. I removed some of the automated commits that aren't robust enough (e.g. some to `Kernel/Sections.h` and `Kernel/Arch/` in general). I squashed them together based on directory. I reset authorship by running `git rebase -i`, and putting `x git commit --amend --author="foo <bar>" --no-edit` after each commit (different author here to prevent copy-paste errors).

This PR is currently in "draft" mode to signal that I'm still ironing out some issues with x68_64, and once I'm absolutely certain that it should pass CI I'll try again.

BenWiederhake commented:
Changes since last push:
- Rebased to avoid conflict
- Fix breakage around `Region::physical_page` by removing commits that touch it (`Region::physical_page` is defined as inline in `VMObject.h`, which includes and depends on `Region.h`, so `Region.h can't include `VMObject.h`, but if someone ends up including only `Region.h` the result is a weird linker error. Aaargh!)
- Add in all includes of `AK/Platform.h` because there were quite a few hidden usages.
- I rebuilt on i686 and x86_64, so I really expect it to work now.
- It doesn't build for aarch64, but that seems to be unrelated.

BenWiederhake commented:
This is just evil.

Apparently LibJS can silently become buggy: "[a missing LibJS include caused the wrong JS::Value constructor to be chosen](https://discord.com/channels/830522505605283862/830739873119207426/886696291361501185)".

Here's the diff that adds back in the includes that are not used but mysteriously are very important (because they cause wrong results at runtime when left out):

```diff
diff --git a/Userland/Libraries/LibJS/Runtime/PropertyDescriptor.cpp b/Userland/Libraries/LibJS/Runtime/PropertyDescriptor.cpp
index 32a520e0fc..dfdaeae9cc 100644
--- a/Userland/Libraries/LibJS/Runtime/PropertyDescriptor.cpp
+++ b/Userland/Libraries/LibJS/Runtime/PropertyDescriptor.cpp
@@ -5,6 +5,7 @@
  */
 
 #include <LibJS/Runtime/Error.h>
+#include <LibJS/Runtime/FunctionObject.h>
 #include <LibJS/Runtime/GlobalObject.h>
 #include <LibJS/Runtime/Object.h>
 #include <LibJS/Runtime/PropertyDescriptor.h>
diff --git a/Userland/Libraries/LibJS/Runtime/Temporal/PlainTimePrototype.cpp b/Userland/Libraries/LibJS/Runtime/Temporal/PlainTimePrototype.cpp
index bbf73bc8dc..78c9b6d535 100644
--- a/Userland/Libraries/LibJS/Runtime/Temporal/PlainTimePrototype.cpp
+++ b/Userland/Libraries/LibJS/Runtime/Temporal/PlainTimePrototype.cpp
@@ -6,6 +6,7 @@
 
 #include <AK/TypeCasts.h>
 #include <LibJS/Runtime/GlobalObject.h>
+#include <LibJS/Runtime/Temporal/Calendar.h>
 #include <LibJS/Runtime/Temporal/PlainDate.h>
 #include <LibJS/Runtime/Temporal/PlainDateTime.h>
 #include <LibJS/Runtime/Temporal/PlainTime.h>
```

So, uh, changes since last push:
- Rebased to avoid conflicts.
- Fixed that silent LibJS bug, see above diff.

BenWiederhake commented:
Changes since last push:
* Rebased due to conflicts
* Removed the Fuzz removals

trflynn89 commented on Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp:
```diff
@@ -29,7 +28,6 @@
 #include <LibJS/Runtime/PropertyName.h>
 #include <LibJS/Runtime/ProxyObject.h>
 #include <LibJS/Runtime/Reference.h>
-#include <LibJS/Runtime/Utf16String.h>
```
This type is used in `get_substitution` - the `replace_string` var is a `Utf16String`

trflynn89 commented on Userland/Libraries/LibJS/Runtime/StringPrototype.cpp:
```diff
@@ -6,16 +6,13 @@
 #include <LibJS/Runtime/Error.h>
 #include <LibJS/Runtime/GlobalObject.h>
 #include <LibJS/Runtime/Intl/AbstractOperations.h>
-#include <LibJS/Runtime/PrimitiveString.h>
```
`PrimitiveString.h` is where `js_string` is defined, which is used a bunch in this file. I spot checked a couple other files where this header is removed and they also use `js_string` (e.g. `SymbolPrototype.cpp` right below here)

trflynn89 commented on Userland/Libraries/LibJS/Runtime/Temporal/CalendarPrototype.cpp:
```diff
@@ -8,7 +8,6 @@
 #include <LibJS/Runtime/Array.h>
 #include <LibJS/Runtime/GlobalObject.h>
 #include <LibJS/Runtime/IteratorOperations.h>
-#include <LibJS/Runtime/Temporal/AbstractOperations.h>
```
I didn't check every diff, but this header is used by every file it was removed from that I did check - see `get_options_object`

trflynn89 commented:
My two cents: maybe this PR should be broken up into multiple, smaller PRs that can actually be reviewed. The script seems to have some false positives, so doing these changes in smaller PRs seems better to make sure we aren't removing used includes.

BenWiederhake commented:
I'm afraid I agree.

BenWiederhake commented:
In particular, for the next attempt I'll use LibCpp, which can more easily check whether any accessed type is defined in a header.

bgianfo commented:
@BenWiederhake  did you see: https://github.com/SerenityOS/serenity/pull/9342 ?

Seems like a much more sustainable approach.

BenWiederhake commented:
The few times I tried to use iwyu, it never worked properly for me. I'm not attempting that approach, but you're welcome to do it and prove me wrong :D
